typescript/sdk/src/core/HyperlaneCore.ts

```diff
@@ -85,6 +89,50 @@ export class HyperlaneCore extends HyperlaneApp<CoreFactories> {
     return this.multiProvider.getChainName(message.parsed.destination);
   }
 
+  getRecipientIsmAddress(message: DispatchedMessage): Promise<Address> {
+    const destinationMailbox = this.contractsMap[this.getDestination(message)];
+    const ethAddress = bytes32ToAddress(message.parsed.recipient);
+    return destinationMailbox.mailbox.recipientIsm(ethAddress);
+  }
+
+  async getRecipientIsmConfig(
+    message: DispatchedMessage,
+  ): Promise<WithAddress<IsmConfig>> {
+    const destinationChain = this.getDestination(message);
+    const ismReader = new EvmIsmReader(this.multiProvider, destinationChain);
+    const address = await this.getRecipientIsmAddress(message);
+    return ismReader.deriveIsmConfig(address);
+  }
+
+  async relayMessage(
+    message: DispatchedMessage,
+  ): Promise<ethers.ContractReceipt> {
+    const destinationChain = this.getDestination(message);
+    const mailbox = this.contractsMap[destinationChain].mailbox;
+    const ismConfig = await this.getRecipientIsmConfig(message);
+    if (typeof ismConfig === 'string') {
+      throw new Error(
+        `No ISM config found for recipient ${message.parsed.recipient}`,
+      );
+    }
+
+    // TODO: implement metadata builders
+    let metadata: string;
+    const moduleType = ismTypeToModuleType(ismConfig.type);
+    switch (moduleType) {
+      case ModuleType.NULL:
+        metadata = '0x';
+        break;
+      default:
+        throw new Error(`Unsupported module type ${moduleType}`);
+    }
```

Comment:
In my ongoing war against switch statements, the worst control flow tool ever invented, I'll complain one more time:
is this really better than
```ts
if ( moduleType === ModuelType.NULL) metadata = '0x';
else throw new Error(`Unsupported module type ${moduleType}`);
```

Reply:
+1, so long as `ModuleType.NULL` comes first in the if statement (purely cosmetic)

LOGAF super low on switch statements

Reply:
rossy v switch statements
![image](https://i.pinimg.com/originals/01/c3/6a/01c36acb87daa7334464741c6f2e7b03.gif)

Reply:
I got rid of the mutation but keeping the switch case after chatting offline :P